pkg/kubelet/cm/helpers_unsupported.go

```diff
@@ -46,7 +46,7 @@ func MilliCPUToShares(milliCPU int64) int64 {
 }
 
 // ResourceConfigForPod takes the input pod and outputs the cgroup resource config.
-func ResourceConfigForPod(pod *v1.Pod, enforceCPULimit bool, cpuPeriod uint64, enforceMemoryQoS bool) *ResourceConfig {
+func ResourceConfigForPod(pod *v1.Pod, enforceCPULimit bool, cpuPeriod uint64, enforceMemoryQoS bool, cpuManagerPolicy string) *ResourceConfig {
```

Comment:
if we use the API `PodContainsPinnedCpus` (or something like it), we don't need a new parameter, and we have less coupling anyway.

Reply:
hmm one issue with relying on `PodContainsPinnedCpus` in cpu_manager is that ResourceConfigForPod as a function does not have access to the cpumanager state. If we go down that path then `ResourceConfigForPod` will sometimes return know-to-be-incorrect ResourceConfigs. That's not intuitive and will surely lead to errors down the road. I think it's cleaner if `ResourceConfigForPod` always returns what we expect the correct `ResourceConfig` to be given the current kubelet settings. 

I like the idea of a method `PodContainsPinnedCpus`, but we also need a way to determine prior to assignment if a pod *should* be allocated a pinned cpu so that the cgroup can be configured correctly from the get-go for both containers/sandbox etc

Reply:
Yeah, that was the main implementation issue I had. So my implementation did this at runtime after the pinned container started. That was the "best" I could do to satisfy the request.

Reply:
I'm still evaluating options. I agree about setting correct values from the beginning is strongly preferred.

Reply:
Resuming this issue. I'm about to open the code, but if we are at the point on which we are creating the pod sandbox (for which we need the resource config), then the pod admission completed, then the cpumanager state is already set, then calling `PodContainsPinnedCpus` should return the correct thing. Anything I'm missing from this picture?

Reply:
The following seems a promising way forward:
```
diff --git a/pkg/kubelet/cm/container_manager.go b/pkg/kubelet/cm/container_manager.go
index 70edb484109..4d7b8b6cd63 100644
--- a/pkg/kubelet/cm/container_manager.go
+++ b/pkg/kubelet/cm/container_manager.go
@@ -146,6 +146,9 @@ type ContainerManager interface {
 	// Updates returns a channel that receives an Update when the device changed its status.
 	Updates() <-chan resourceupdates.Update
 
+	// TODO
+	HasExclusiveCPUs(pod *v1.Pod) bool
+
 	// Implements the PodResources Provider API
 	podresources.CPUsProvider
 	podresources.DevicesProvider
diff --git a/pkg/kubelet/cm/container_manager_linux.go b/pkg/kubelet/cm/container_manager_linux.go
index 78d3366750e..d09f03acd02 100644
--- a/pkg/kubelet/cm/container_manager_linux.go
+++ b/pkg/kubelet/cm/container_manager_linux.go
@@ -363,7 +363,7 @@ func (cm *containerManagerImpl) NewPodContainerManager() PodContainerManager {
 			// cpuCFSQuotaPeriod is in microseconds. NodeConfig.CPUCFSQuotaPeriod is time.Duration (measured in nano seconds).
 			// Convert (cm.CPUCFSQuotaPeriod) [nanoseconds] / time.Microsecond (1000) to get cpuCFSQuotaPeriod in microseconds.
 			cpuCFSQuotaPeriod: uint64(cm.CPUCFSQuotaPeriod / time.Microsecond),
-			cpuManagerPolicy:  cm.GetNodeConfig().CPUManagerPolicy,
+			podCPUProvider:    cm,
 		}
 	}
 	return &podContainerManagerNoop{
@@ -371,6 +371,19 @@ func (cm *containerManagerImpl) NewPodContainerManager() PodContainerManager {
 	}
 }
 
+func (cm *containerManagerImpl) HasExclusiveCPUs(pod *v1.Pod) bool {
+	for _, container := range pod.Spec.Containers {
+		exclusiveCPUs := cm.cpuManager.GetExclusiveCPUs(string(pod.UID), container.Name)
+		if !exclusiveCPUs.IsEmpty() {
+			klog.V(4).InfoS("Pod contains container with pinned cpus", "podName", pod.Name, "containerName", container.Name)
+			return true
+		}
+	}
+
+	klog.V(4).InfoS("Pod contains no container with pinned cpus", "podName", pod.Name)
+	return false
+}
+
 func (cm *containerManagerImpl) InternalContainerLifecycle() InternalContainerLifecycle {
 	return &internalContainerLifecycleImpl{cm.cpuManager, cm.memoryManager, cm.topologyManager}
 }
diff --git a/pkg/kubelet/cm/container_manager_stub.go b/pkg/kubelet/cm/container_manager_stub.go
index 35aeee36724..b796e5dc263 100644
--- a/pkg/kubelet/cm/container_manager_stub.go
+++ b/pkg/kubelet/cm/container_manager_stub.go
@@ -190,6 +190,10 @@ func (cm *containerManagerStub) Updates() <-chan resourceupdates.Update {
 	return nil
 }
 
+func (cm *containerManagerStub) HasExclusiveCPUs(pod *v1.Pod) bool {
+	return false
+}
+
 func NewStubContainerManager() ContainerManager {
 	return &containerManagerStub{shouldResetExtendedResourceCapacity: false}
 }
diff --git a/pkg/kubelet/cm/container_manager_windows.go b/pkg/kubelet/cm/container_manager_windows.go
index 232d121223a..0d825e6b2e5 100644
--- a/pkg/kubelet/cm/container_manager_windows.go
+++ b/pkg/kubelet/cm/container_manager_windows.go
@@ -285,3 +285,7 @@ func (cm *containerManagerImpl) UnprepareDynamicResources(ctx context.Context, p
 func (cm *containerManagerImpl) PodMightNeedToUnprepareResources(UID types.UID) bool {
 	return false
 }
+
+func (cm *containerManagerImpl) HasExclusiveCPUs(pod *v1.Pod) bool {
+	return false
+}
diff --git a/pkg/kubelet/cm/fake_container_manager.go b/pkg/kubelet/cm/fake_container_manager.go
index f3f6c6eb509..3abe9c842a7 100644
--- a/pkg/kubelet/cm/fake_container_manager.go
+++ b/pkg/kubelet/cm/fake_container_manager.go
@@ -260,3 +260,7 @@ func (cm *FakeContainerManager) UpdateAllocatedResourcesStatus(pod *v1.Pod, stat
 func (cm *FakeContainerManager) Updates() <-chan resourceupdates.Update {
 	return nil
 }
+
+func (cm *FakeContainerManager) HasExclusiveCPUs(pod *v1.Pod) bool {
+	return false
+}
diff --git a/pkg/kubelet/cm/helpers_linux.go b/pkg/kubelet/cm/helpers_linux.go
index bd564d1398d..ebd19cc695b 100644
--- a/pkg/kubelet/cm/helpers_linux.go
+++ b/pkg/kubelet/cm/helpers_linux.go
@@ -25,10 +25,8 @@ import (
 
 	libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
 	v1 "k8s.io/api/core/v1"
-	apiresource "k8s.io/apimachinery/pkg/api/resource"
 	"k8s.io/apimachinery/pkg/types"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
-	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager"
 
 	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
 	"k8s.io/kubernetes/pkg/api/v1/resource"
@@ -120,7 +118,7 @@ func HugePageLimits(resourceList v1.ResourceList) map[int64]int64 {
 }
 
 // ResourceConfigForPod takes the input pod and outputs the cgroup resource config.
-func ResourceConfigForPod(pod *v1.Pod, enforceCPULimits bool, cpuPeriod uint64, enforceMemoryQoS bool, cpuManagerPolicy string) *ResourceConfig {
+func ResourceConfigForPod(pod *v1.Pod, enforceCPULimits bool, cpuPeriod uint64, enforceMemoryQoS bool) *ResourceConfig {
 	inPlacePodVerticalScalingEnabled := utilfeature.DefaultFeatureGate.Enabled(kubefeatures.InPlacePodVerticalScaling)
 	// sum requests and limits.
 	reqs := resource.PodRequests(pod, resource.PodResourcesOptions{
@@ -147,10 +145,8 @@ func ResourceConfigForPod(pod *v1.Pod, enforceCPULimits bool, cpuPeriod uint64,
 	cpuRequests := int64(0)
 	cpuLimits := int64(0)
 	memoryLimits := int64(0)
-	var cpuRequestQuantity apiresource.Quantity
 
 	if request, found := reqs[v1.ResourceCPU]; found {
-		cpuRequestQuantity = request
 		cpuRequests = request.MilliValue()
 	}
 	if limit, found := limits[v1.ResourceCPU]; found {
@@ -166,11 +162,9 @@ func ResourceConfigForPod(pod *v1.Pod, enforceCPULimits bool, cpuPeriod uint64,
 
 	qosClass := v1qos.GetPodQOS(pod)
 
-	StaticCPUPolicyConditionsSatisfied := cpumanager.StaticCPUPolicyConditionsSatisfied(cpuManagerPolicy, qosClass, &cpuRequestQuantity)
-
 	// quota is not capped when cfs quota is disabled,
 	// or if static cpu policy conditions are satisfied
-	if !enforceCPULimits || StaticCPUPolicyConditionsSatisfied {
+	if !enforceCPULimits {
 		cpuQuota = int64(-1)
 	}
 
diff --git a/pkg/kubelet/cm/helpers_linux_test.go b/pkg/kubelet/cm/helpers_linux_test.go
index b6ea8129b07..df36e4ade79 100644
--- a/pkg/kubelet/cm/helpers_linux_test.go
+++ b/pkg/kubelet/cm/helpers_linux_test.go
@@ -357,7 +357,7 @@ func TestResourceConfigForPod(t *testing.T) {
 
 	for testName, testCase := range testCases {
 
-		actual := ResourceConfigForPod(testCase.pod, testCase.enforceCPULimits, testCase.quotaPeriod, false, testCase.cmPolicy)
+		actual := ResourceConfigForPod(testCase.pod, testCase.enforceCPULimits, testCase.quotaPeriod, false)
 
 		if !reflect.DeepEqual(actual.CPUPeriod, testCase.expected.CPUPeriod) {
 			t.Errorf("unexpected result, test: %v, cpu period not as expected. Expected: %v, Actual:%v", testName, *testCase.expected.CPUPeriod, *actual.CPUPeriod)
diff --git a/pkg/kubelet/cm/helpers_unsupported.go b/pkg/kubelet/cm/helpers_unsupported.go
index c22d384b7ba..526dd8fd7db 100644
--- a/pkg/kubelet/cm/helpers_unsupported.go
+++ b/pkg/kubelet/cm/helpers_unsupported.go
@@ -20,7 +20,7 @@ limitations under the License.
 package cm
 
 import (
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 )
 
@@ -46,7 +46,7 @@ func MilliCPUToShares(milliCPU int64) int64 {
 }
 
 // ResourceConfigForPod takes the input pod and outputs the cgroup resource config.
-func ResourceConfigForPod(pod *v1.Pod, enforceCPULimit bool, cpuPeriod uint64, enforceMemoryQoS bool, cpuManagerPolicy string) *ResourceConfig {
+func ResourceConfigForPod(pod *v1.Pod, enforceCPULimit bool, cpuPeriod uint64, enforceMemoryQoS bool) *ResourceConfig {
 	return nil
 }
 
diff --git a/pkg/kubelet/cm/pod_container_manager_linux.go b/pkg/kubelet/cm/pod_container_manager_linux.go
index 69df8a5d779..541343aa6d7 100644
--- a/pkg/kubelet/cm/pod_container_manager_linux.go
+++ b/pkg/kubelet/cm/pod_container_manager_linux.go
@@ -37,6 +37,11 @@ const (
 	podCgroupNamePrefix = "pod"
 )
 
+// name TBD
+type podCPUProvider interface {
+	HasExclusiveCPUs(pod *v1.Pod) bool
+}
+
 // podContainerManagerImpl implements podContainerManager interface.
 // It is the general implementation which allows pod level container
 // management if qos Cgroup is enabled.
@@ -55,8 +60,8 @@ type podContainerManagerImpl struct {
 	// cpuCFSQuotaPeriod is the cfs period value, cfs_period_us, setting per
 	// node for all containers in usec
 	cpuCFSQuotaPeriod uint64
-	// cpuManagerPolicy is the cpu manager policy
-	cpuManagerPolicy string
+	// TODO
+	podCPUProvider podCPUProvider
 }
 
 // Make sure that podContainerManagerImpl implements the PodContainerManager interface
@@ -74,6 +79,10 @@ func (m *podContainerManagerImpl) Exists(pod *v1.Pod) bool {
 func (m *podContainerManagerImpl) EnsureExists(pod *v1.Pod) error {
 	// check if container already exist
 	alreadyExists := m.Exists(pod)
+	enforceCPULimits := m.enforceCPULimits
+	if m.podCPUProvider.HasExclusiveCPUs(pod) {
+		enforceCPULimits = false
+	}
 	if !alreadyExists {
 		enforceMemoryQoS := false
 		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.MemoryQoS) &&
@@ -84,7 +93,7 @@ func (m *podContainerManagerImpl) EnsureExists(pod *v1.Pod) error {
 		podContainerName, _ := m.GetPodContainerName(pod)
 		containerConfig := &CgroupConfig{
 			Name:               podContainerName,
-			ResourceParameters: ResourceConfigForPod(pod, m.enforceCPULimits, m.cpuCFSQuotaPeriod, enforceMemoryQoS, m.cpuManagerPolicy),
+			ResourceParameters: ResourceConfigForPod(pod, enforceCPULimits, m.cpuCFSQuotaPeriod, enforceMemoryQoS),
 		}
 		if m.podPidsLimit > 0 {
 			containerConfig.ResourceParameters.PidsLimit = &m.podPidsLimit
diff --git a/pkg/kubelet/kuberuntime/kuberuntime_manager.go b/pkg/kubelet/kuberuntime/kuberuntime_manager.go
index 86f96565c3b..2f045335ab0 100644
--- a/pkg/kubelet/kuberuntime/kuberuntime_manager.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_manager.go
@@ -665,7 +665,11 @@ func (m *kubeGenericRuntimeManager) computePodResizeAction(pod *v1.Pod, containe
 func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *kubecontainer.PodStatus, podContainerChanges podActions, result kubecontainer.PodSyncResult) {
 	pcm := m.containerManager.NewPodContainerManager()
 	//TODO(vinaykul,InPlacePodVerticalScaling): Figure out best way to get enforceMemoryQoS value (parameter #4 below) in platform-agnostic way
-	podResources := cm.ResourceConfigForPod(pod, m.cpuCFSQuota, uint64((m.cpuCFSQuotaPeriod.Duration)/time.Microsecond), false, m.containerManager.GetNodeConfig().CPUManagerPolicy)
+	enforceCPULimits := m.cpuCFSQuota
+	if m.containerManager.HasExclusiveCPUs(pod) {
+		enforceCPULimits = false
+	}
+	podResources := cm.ResourceConfigForPod(pod, enforceCPULimits, uint64((m.cpuCFSQuotaPeriod.Duration)/time.Microsecond), false)
 	if podResources == nil {
 		klog.ErrorS(nil, "Unable to get resource configuration", "pod", pod.Name)
 		result.Fail(fmt.Errorf("Unable to get resource configuration processing resize for pod %s", pod.Name))
diff --git a/pkg/volume/emptydir/empty_dir.go b/pkg/volume/emptydir/empty_dir.go
index 83d62977635..1d685ae83da 100644
--- a/pkg/volume/emptydir/empty_dir.go
+++ b/pkg/volume/emptydir/empty_dir.go
@@ -18,10 +18,11 @@ package emptydir
 
 import (
 	"fmt"
-	"k8s.io/kubernetes/pkg/kubelet/util/swap"
 	"os"
 	"path/filepath"
 
+	"k8s.io/kubernetes/pkg/kubelet/util/swap"
+
 	"k8s.io/klog/v2"
 	"k8s.io/mount-utils"
 	utilstrings "k8s.io/utils/strings"
@@ -123,7 +124,7 @@ func calculateEmptyDirMemorySize(nodeAllocatableMemory *resource.Quantity, spec
 	// determine pod resource allocation
 	// we use the same function for pod cgroup assignment to maintain consistent behavior
 	// NOTE: this could be nil on systems that do not support pod memory containment (i.e. windows)
-	podResourceConfig := cm.ResourceConfigForPod(pod, false, uint64(100000), false, "none")
+	podResourceConfig := cm.ResourceConfigForPod(pod, false, uint64(100000), false)
 	if podResourceConfig != nil && podResourceConfig.Memory != nil {
 		podMemoryLimit := resource.NewQuantity(*(podResourceConfig.Memory), resource.BinarySI)
 		// ensure 0 < value < size
```

Reply:
in any case, we need to add the new feature gate

Reply:
adding a sketch with a feature gate; needs more work, but this is the general idea
```
diff --git a/pkg/features/kube_features.go b/pkg/features/kube_features.go
index 11fb8d88055..8346db18792 100644
--- a/pkg/features/kube_features.go
+++ b/pkg/features/kube_features.go
@@ -938,6 +938,12 @@ const (
 	//
 	// Enables the image volume source.
 	ImageVolume featuregate.Feature = "ImageVolume"
+
+	// owner: @ffromani
+	// beta: v1.31
+	//
+	// Disable CPU Quota for pods which have exclusive CPUs allocated
+	DisableCPUQuotaWithExclusiveCPUs featuregate.Feature = "DisableCPUQuotaWithExclusiveCPUs"
 )
 
 func init() {
diff --git a/pkg/features/versioned_kube_features.go b/pkg/features/versioned_kube_features.go
index 6398233b835..2bee975f01f 100644
--- a/pkg/features/versioned_kube_features.go
+++ b/pkg/features/versioned_kube_features.go
@@ -435,4 +435,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	KubeletRegistrationGetOnExistsOnly: {
 		{Version: version.MustParse("1.32"), Default: false, PreRelease: featuregate.Deprecated},
 	},
+	DisableCPUQuotaWithExclusiveCPUs: {
+		{Version: version.MustParse("1.32"), Default: true, PreRelease: featuregate.Beta},
+	},
 }
diff --git a/pkg/kubelet/cm/pod_container_manager_linux.go b/pkg/kubelet/cm/pod_container_manager_linux.go
index 541343aa6d7..af541d50f35 100644
--- a/pkg/kubelet/cm/pod_container_manager_linux.go
+++ b/pkg/kubelet/cm/pod_container_manager_linux.go
@@ -80,8 +80,10 @@ func (m *podContainerManagerImpl) EnsureExists(pod *v1.Pod) error {
 	// check if container already exist
 	alreadyExists := m.Exists(pod)
 	enforceCPULimits := m.enforceCPULimits
-	if m.podCPUProvider.HasExclusiveCPUs(pod) {
-		enforceCPULimits = false
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DisableCPUQuotaWithExclusiveCPUs) {
+		if m.podCPUProvider.HasExclusiveCPUs(pod) {
+			enforceCPULimits = false
+		}
 	}
 	if !alreadyExists {
 		enforceMemoryQoS := false
diff --git a/pkg/kubelet/kuberuntime/kuberuntime_manager.go b/pkg/kubelet/kuberuntime/kuberuntime_manager.go
index 2f045335ab0..f3216f8111a 100644
--- a/pkg/kubelet/kuberuntime/kuberuntime_manager.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_manager.go
@@ -51,6 +51,7 @@ import (
 	"k8s.io/kubernetes/pkg/credentialprovider"
 	"k8s.io/kubernetes/pkg/credentialprovider/plugin"
 	"k8s.io/kubernetes/pkg/features"
+	kubefeatures "k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/kubelet/cm"
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 	"k8s.io/kubernetes/pkg/kubelet/events"
@@ -666,8 +667,10 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
 	pcm := m.containerManager.NewPodContainerManager()
 	//TODO(vinaykul,InPlacePodVerticalScaling): Figure out best way to get enforceMemoryQoS value (parameter #4 below) in platform-agnostic way
 	enforceCPULimits := m.cpuCFSQuota
-	if m.containerManager.HasExclusiveCPUs(pod) {
-		enforceCPULimits = false
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DisableCPUQuotaWithExclusiveCPUs) {
+		if m.containerManager.HasExclusiveCPUs(pod) {
+			enforceCPULimits = false
+		}
 	}
 	podResources := cm.ResourceConfigForPod(pod, enforceCPULimits, uint64((m.cpuCFSQuotaPeriod.Duration)/time.Microsecond), false)
 	if podResources == nil {
```

Reply:
@ffromani done.

added feature gate, incorporated above changes to surface pinned cpus via new interface method. 

that takes care of all calls from within `ResourceConfigForPod`, which can remain untouched by this PR.

We still need some way to determine if the cfsQuota should be removed inside `calculateLinuxResources` as that's called from a number of other places. 

I've kept the `StaticCPUPolicyConditionsSatisfied` method (takes qos, cpu req, policyname), locked behind feature gate. I also swapped to use the same computation logic as the static policy so we're consistent in how we determine if a cpu request satsifies the policy constraints:

```
cpuRequest.Value()*1000 == cpuRequest.MilliValue()
```

let me know what you think